src/v/security/gssapi_authenticator.cc

```diff
@@ -0,0 +1,514 @@
+ *      min_stat        the GSS-API minor status code
+ *
+ * Effects:
+ *
+ * The GSS-API messages associated with maj_stat and min_stat are
+ * displayed on stderr, each preceeded by "GSS-API error <msg>:
+" and
+ * followed by a newline.
+ */
+void display_status(const char* msg, OM_uint32 maj_stat, OM_uint32 min_stat) {
+    display_status_1(msg, maj_stat, GSS_C_GSS_CODE);
+    display_status_1(msg, min_stat, GSS_C_MECH_CODE);
+}
+
+ss::future<result<bytes>>
+gssapi_authenticator::authenticate(bytes_view auth_bytes) {
+    vlog(seclog.info, "gss {} authenticate", _state);
+    switch (_state) {
+    case state::init: {
+        auto principal = config::shard_local_cfg().sasl_kerberos_principal();
+        auto keytab = config::shard_local_cfg().sasl_kerberos_keytab();
+
+        if (auto res = co_await _worker.submit(
+              [this, principal, keytab]() { return init(principal, keytab); });
+            res.has_error()) {
+            vlog(seclog.info, "gss {} authenticate failed", _state);
+            _state = state::failed;
+            co_return res.assume_error();
+        }
+    }
+        [[fallthrough]];
+    case state::more: {
+        auto res = co_await _worker.submit(
+          [this, auth_bytes]() { return more(auth_bytes); });
+        if (res.has_error()) {
+            vlog(seclog.info, "gss {} authenticate failed", _state);
+            _state = state::failed;
+        }
+        co_return res;
+    }
+    case state::ssfcap: {
+        auto res = co_await _worker.submit(
+          [this, auth_bytes]() { return ssfcap(auth_bytes); });
+        if (res.has_error()) {
+            vlog(seclog.info, "gss {} authenticate failed", _state);
+            _state = state::failed;
+        }
+        co_return res;
+    }
+    case state::ssfreq: {
+        auto res = co_await _worker.submit(
+          [this, auth_bytes]() { return ssfreq(auth_bytes); });
+        if (res.has_error()) {
+            vlog(seclog.info, "gss {} authenticate failed", _state);
+            _state = state::failed;
+        }
+        co_return res;
+    }
+    case state::complete:
+    case state::failed:
+        break;
+    }
+    vlog(seclog.info, "gss {} authenticate failed", _state);
+    _state = state::failed;
+    co_return errc::invalid_scram_state;
+}
+
+result<void>
+gssapi_authenticator::init(ss::sstring principal, ss::sstring keytab) {
+    OM_uint32 minor_status{};
+    OM_uint32 major_status{};
+
+    gss::buffer_view service_name{principal};
+    gss::name server_name{};
+
+    major_status = ::gss_import_name(
+      &minor_status, &service_name, GSS_C_NT_HOSTBASED_SERVICE, &server_name);
+
+    vlog(
+      seclog.trace,
+      "gss {} import name {}:{} - {}",
+      _state,
+      major_status,
+      minor_status,
+      principal);
+    if (major_status != GSS_S_COMPLETE) {
+        display_status("import name", major_status, minor_status);
+        _state = state::failed;
+        return errc::invalid_scram_state;
```

Comment:
Probably a tbd but should these return codes be renamed (namely since this isn't scram that's causing the error)?

Reply:
Oh, good, point. Noted.

Reply:
Done